### PR TITLE
Add setup rancher and harvester in air-gapped environment

### PR DESCRIPTION
**Pull request by FrankYang0529**

### CODE REVIEW

**Review by TachunLin — APPROVED** [review]
LGTM. Appreciate for the update

**Comment by guangbochen** [comment]
The content looks good to me, but in regards to the Rancher air-gap setup, I would prefer to link it to a particular Rancher doc page as Rancher should constantly maintain and update it, but we can always add some important notes on this section.

File: docs/airgap.md [hunk 1]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
 The Harvester ISO image contains all the packages to make it work in an air gapped environment.
 
+## Working without HTTP Proxy
+
+In this example, we use KVM to provision Harvester and integrate it with Rancher.
+
+### Setup offline Harvester
+
+1. Clone [harvester/ipxe-examples](https://github.com/harvester/ipxe-examples).
+1. Go to `vagrant-pxe-harvester` folder.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:22]:
```suggestion
1. Select the `vagrant-pxe-harvester` folder.
```

File: docs/airgap.md [hunk 2]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+1. Edit [settings.yml](https://github.com/harvester/ipxe-examples/blob/c8267b6270660255bf71149a1fef3a7a914550de/vagrant-pxe-harvester/settings.yml#L44) and set `offline: true`.
+1. Execute [setup_harvester.sh](https://github.com/harvester/ipxe-examples/blob/main/vagrant-pxe-harvester/setup_harvester.sh) script.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:24]:
```suggestion
1. Execute the [setup_harvester.sh](https://github.com/harvester/ipxe-examples/blob/main/vagrant-pxe-harvester/setup_harvester.sh) script.
```

File: docs/airgap.md [hunk 3]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+
+### Deploy Rancher on k3s and setup private registry in another VM.
+
+1. Create another KVM with two network interfaces `harvester` and `vagrant-libvirt`. `harvester` network interface is for intranet and `vagrant-libvirt` is for internet. We need `vagrant-libvirt` to download all required resources. We will remove it before we start Rancher. Also, please give this VM at least 300GB to save all required images.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:28]:
```suggestion
1. Create another KVM with two network interfaces, `harvester` and `vagrant-libvirt`. 
- `harvester` is for intranet and `vagrant-libvirt` is for internet. 
- We need `vagrant-libvirt` to download all required resources. We will remove it before we start Rancher. 
- Configure this VM with at least 300GB to save all required images.
```

File: docs/airgap.md [hunk 4]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+In this example, we use KVM to provision Harvester and integrate it with Rancher.
+
+### Setup offline Harvester
+
+1. Clone [harvester/ipxe-examples](https://github.com/harvester/ipxe-examples).
+1. Go to `vagrant-pxe-harvester` folder.
+1. Edit [settings.yml](https://github.com/harvester/ipxe-examples/blob/c8267b6270660255bf71149a1fef3a7a914550de/vagrant-pxe-harvester/settings.yml#L44) and set `offline: true`.
+1. Execute [setup_harvester.sh](https://github.com/harvester/ipxe-examples/blob/main/vagrant-pxe-harvester/setup_harvester.sh) script.
+
+### Deploy Rancher on k3s and setup private registry in another VM.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:26]:
```suggestion
### Deploy Rancher on K3s and setup private registry in another VM
```

File: docs/airgap.md [hunk 5]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+
+1. Create another KVM with two network interfaces `harvester` and `vagrant-libvirt`. `harvester` network interface is for intranet and `vagrant-libvirt` is for internet. We need `vagrant-libvirt` to download all required resources. We will remove it before we start Rancher. Also, please give this VM at least 300GB to save all required images.
+1. Install [Docker](https://www.docker.com/) and [Helm](https://helm.sh/).
+1. Create `certs` folder.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:30]:
```suggestion
1. Create a `certs` folder.
```

File: docs/airgap.md [hunk 6]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+  -newkey rsa:4096 -nodes -sha256 -keyout certs/domain.key \
+  -addext "subjectAltName = DNS:myregistry.local" \
+  -x509 -days 365 -out certs/domain.crt
+```
+1. Move certificate files to `/etc/docker/certs.d`.
+```
+sudo mkdir -p /etc/docker/certs.d/myregistry.local:5000
+sudo cp certs/domain.crt /etc/docker/certs.d/myregistry.local:5000/domain.crt
+```
+1. Start private registry.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:46]:
```suggestion
1. Start the private registry.
```

File: docs/airgap.md [hunk 7]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+  -p 5000:5000 \
+  --restart=always \
+  --name registry \
+  -v "$(pwd)"/certs:/certs \
+  -v "$(pwd)"/registry:/var/lib/registry \
+  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
+  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
+  registry:2
+```
+1. Add `myregistry.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private ip.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:58]:
```suggestion
1. Add `myregistry.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private IP.
```

File: docs/airgap.md [hunk 8]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+    echo 'This requires you to pass a version for the url like "v2.6.4"'
+    exit 1
+fi
+wget https://github.com/rancher/rancher/releases/download/$1/rancher-images.txt
+wget https://github.com/rancher/rancher/releases/download/$1/rancher-load-images.sh
+wget https://github.com/rancher/rancher/releases/download/$1/rancher-save-images.sh
+chmod +x ./rancher-save-images.sh
+chmod +x ./rancher-load-images.sh
+```
+1. Make `get-rancher` script be excutable.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:77]:
```suggestion
1. Make the `get-rancher` script executable.
```

File: docs/airgap.md [hunk 9]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+  -e REGISTRY_HTTP_TLS_CERTIFICATE=/certs/domain.crt \
+  -e REGISTRY_HTTP_TLS_KEY=/certs/domain.key \
+  registry:2
+```
+1. Add `myregistry.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private ip.
+```
+# vim /etc/hosts
+192.168.0.50 myregistry.local
+```
+1. Create `get-rancher` script.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:63]:
```suggestion
1. Create a `get-rancher` script.
```

File: docs/airgap.md [hunk 10]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+helm repo add jetstack https://charts.jetstack.io/
+helm repo update
+helm fetch jetstack/cert-manager --version v1.7.1
+helm template ./cert-manager-v1.7.1.tgz | awk '$1 ~ /image:/ {print $2}' | sed s/\"//g >> ./rancher-images.txt
+```
+1. Sort `rancher-images.txt`.
+```
+sort -u rancher-images.txt -o rancher-images.txt
+```
+1. Get images. This step may take 1 to 2 hours. It depends on your network speed.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:96]:
```suggestion
1. Get images. This step may take 1 to 2 hours depending on your network speed.
```

File: docs/airgap.md [hunk 11]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+./rancher-save-images.sh --image-list ./rancher-images.txt
+```
+1. Load images to local registry.
+```
+./rancher-load-images.sh --image-list ./rancher-images.txt --registry myregistry.local:5000
+```
+1. Create `get-k3s` script.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:104]:
```suggestion
1. Create a `get-k3s` script.
```

File: docs/airgap.md [hunk 12]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+    echo 'This requires you to pass a version for the url like "v1.23.4+k3s1"'
+    exit 1
+fi
+wget https://github.com/k3s-io/k3s/releases/download/$1/k3s-airgap-images-amd64.tar
+wget https://github.com/k3s-io/k3s/releases/download/$1/k3s
+wget https://get.k3s.io/ -O install.sh
+chmod +x ./k3s
+chmod +x ./install.sh
+```
+1. Make `get-k3s` be excutable.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:118]:
```suggestion
1. Make `get-k3s` excuteable.
```

File: docs/airgap.md [hunk 13]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+helm repo add rancher-latest https://releases.rancher.com/server-charts/latest
+helm fetch rancher-latest/rancher --version=v2.6.4
+```
+1. Download `cert-manager-crds.yaml`.
+```
+mkdir cert-manager
+curl -L -o cert-manager/cert-manager-crd.yaml https://github.com/jetstack/cert-manager/releases/download/v1.7.1/cert-manager.crds.yaml
+```
+1. Remove `vagrant-libvirt` network interface. We've downloaded all required resources. We can cut network from this step.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:136]:
```suggestion
1. Remove the `vagrant-libvirt` network interface. Once the required resources download, you can remove the network.
```

File: docs/airgap.md [hunk 14]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+1. Move `k3s-airgap-images-amd64.tar` to `/var/lib/rancher/k3s/agent/images/`.
+```
+sudo mkdir -p /var/lib/rancher/k3s/agent/images/
+sudo cp k3s-airgap-images-amd64.tar /var/lib/rancher/k3s/agent/images/
+```
+1. Create `/etc/rancher/k3s` folder.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:142]:
```suggestion
1. Create a `/etc/rancher/k3s` folder.
```

File: docs/airgap.md [hunk 15]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+mirrors:
+  docker.io:
+    endpoint:
+      - "https://myregistry.local:5000/"
+configs:
+  "myregistry.local:5000":
+    tls:
+      insecure_skip_verify: true
+```
+1. Install k3s.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:158]:
```suggestion
1. Install K3s.
```

File: docs/airgap.md [hunk 16]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+sudo chown $USER ~/.kube/config
+export KUBECONFIG=~/.kube/config
+```
+1. Install cert-manager.
+```
+kubectl create namespace cert-manager
+kubectl apply -f cert-manager/cert-manager-crd.yaml
+kubectl apply -R -f ./cert-manager
+```
+1. Create CA private key and certifacate file.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:184]:
```suggestion
1. Create a CA private key and a certificate file.
```

File: docs/airgap.md [hunk 17]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+openssl genrsa -out cakey.pem 2048
+openssl req -x509 -sha256 -new -nodes -key cakey.pem -days 3650 -out cacerts.pem -subj "/CN=cattle-ca"
+```
+1. Create `openssl.cnf`. Remember to change `192.168.0.50` to your private ip.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:189]:
```suggestion
1. Create `openssl.cnf`. Remember to change `192.168.0.50` to your private IP
```

File: docs/airgap.md [hunk 18]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+extendedKeyUsage = clientAuth, serverAuth
+subjectAltName = @alt_names
+[alt_names]
+DNS.1 = myrancher.local
+IP.1 = 192.168.0.50
+```
+1. Generate private key and certificate file for `myrancher.local`.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:204]:
```suggestion
1. Generate s private key and a certificate file for `myrancher.local`.
```

File: docs/airgap.md [hunk 19]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+openssl genrsa -out tls.key 2048
+openssl req -sha256 -new -key tls.key -out tls.csr -subj "/CN=myrancher.local" -config openssl.cnf
+openssl x509 -sha256 -req -in tls.csr -CA cacerts.pem \
+    -CAkey cakey.pem -CAcreateserial -out tls.crt \
+    -days 3650 -extensions v3_req \
+    -extfile openssl.cnf
+```
+1. Create `cattle-system` namespace.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:213]:
```suggestion
1. Create a `cattle-system` namespace.
```

File: docs/airgap.md [hunk 20]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+kubectl create ns cattle-system
+```
+1. Create `tls.sa` secret.
+```
+kubectl -n cattle-system create secret generic tls-ca \
+  --from-file=cacerts.pem=./cacerts.pem
+```
+1. Create `tls-rancher-ingress` secret.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:222]:
```suggestion
1. Create a `tls-rancher-ingress` secret.
```

File: docs/airgap.md [hunk 21]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+    --set systemDefaultRegistry=myregistry.local:5000 \
+    --set useBundledSystemChart=true \
+    --set ingress.tls.source=secret \
+    --set privateCA=true
+```
+1. Install Rancher.
+```
+kubectl -n cattle-system apply -R -f ./rancher
+```
+1. Add `myrancher.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private ip.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:245]:
```suggestion
1. Add `myrancher.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private IP.
```

File: docs/airgap.md [hunk 22]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+configs:
+  "myregistry.local:5000":
+    tls:
+      insecure_skip_verify: true
+```
+1. (Harvester VM) Restart RKE2.
+```
+systemctl restart rke2-server.service
+```
+1. (Harvester VM) Update `rke2-coredns-rke2-coredns` ConfigMap. Remember to change `192.168.0.50` to your private ip.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:274]:
```suggestion
1. (Harvester VM) Update the `rke2-coredns-rke2-coredns` ConfigMap. Remember to change `192.168.0.50` to your private IP.
```

File: docs/airgap.md [hunk 23]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+```
+1. Add `myrancher.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private ip.
+```
+# vim /etc/hosts
+192.168.0.50 myregistry.local myrancher.local
+```
+
+### Integrate Harvester with Rancher in air-gapped environment
+
+1. (Harvester VM) Add `myregistry.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private ip.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:253]:
```suggestion
1. (Harvester VM) Add `myregistry.local` to `/etc/hosts`. Remember to change `192.168.0.50` to your private IP.
```

File: docs/airgap.md [hunk 24]
```diff
@@ -12,6 +12,287 @@ This section describes how to use Harvester in an air gapped environment. Some u
...
+  Corefile: ".:53 {\n    errors \n    health  {\n        lameduck 5s\n    }\n    ready
+    \n    kubernetes   cluster.local  cluster.local in-addr.arpa ip6.arpa {\n        pods
+    insecure\n        fallthrough in-addr.arpa ip6.arpa\n        ttl 30\n    }\n    prometheus
+    \  0.0.0.0:9153\n   hosts /etc/coredns/customdomains.db myrancher.local {\n
+    \   fallthrough\n    }\n forward   . /etc/resolv.conf\n    cache   30\n    loop
+    \n    reload \n    loadbalance \n}"
+  customdomains.db: |
+    192.168.0.50 myrancher.local
+```
+1. (Harvester VM) Update `rke2-coredns-rke2-coredns` Deployment.
```
Comment (LucasSaintarbor) [inline docs/airgap.md:287]:
```suggestion
1. (Harvester VM) Update the `rke2-coredns-rke2-coredns` deployment.
```

**Review by LucasSaintarbor — APPROVED** [review]
Made some small changes. LGTM 👍 @TachunLin is it okay to merge this? I will create a different PR to add this to https://github.com/harvester/docs/blob/main/versioned_docs/version-v1.1/airgap.md

**Comment by TachunLin** [comment]
Hi @LucasSaintarbor, thanks for the check. 
It looks good to me. Please feel free to update. 

if you are going to create a separate PR for adding the changes. 
A suggested update is we can add indent to the code block to let the numbering bullet work better. 
image

**Comment by LucasSaintarbor** [comment]
Hi @FrankYang0529, sorry for being late to this. Since this PR is older, is was created using the old site generator [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/) and not [Docusaurus 2](https://docusaurus.io/). I am going to create a new PR from my fork https://github.com/LucasSaintarbor/harvester-docs/tree/main with the same changes, and then I'll follow up here to close this PR.

**Comment by FrankYang0529** [comment]
> Hi @FrankYang0529, sorry for being late to this. Since this PR is older, is was created using the old site generator [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/) and not [Docusaurus 2](https://docusaurus.io/). I am going to create a new PR from my fork https://github.com/LucasSaintarbor/harvester-docs/tree/main with the same changes, and then I'll follow up here to close this PR.

Hi @LucasSaintarbor, thanks for the help. Sounds good to me. Feel free to close this PR after you submit a new one. 👍

**Comment by LucasSaintarbor** [comment]
Per https://github.com/harvester/docs/pull/137#issuecomment-1652842247, opened new PR https://github.com/harvester/docs/pull/371 - Updated airgap.md with steps to setup rancher and harvester in air-gapped env #371

**Pull request closed without merging**